brainiak/isfc.py

```diff
@@ -80,37 +86,50 @@ def isc(D, collapse_subj=True, return_p=False, num_perm=1000,
     n_vox = D.shape[0]
     n_subj = D.shape[2]
 
-    if return_p:
-        n_perm = num_perm
-    else:
-        n_perm = 0
-
-    ISC = np.zeros((n_vox, n_subj, n_perm + 1))
+    n_perm = num_perm*int(return_p)
+    max_null, min_null = gen_null_array(n_subj, n_perm,
+                                        collapse_subj, float_type)
+    ISC = np.zeros((n_vox, n_subj), dtype=float_type)
+
+    for loo_subj in range(n_subj):
+        group = np.mean(D[:, :, np.arange(n_subj) != loo_subj], axis=2)
+        subj = D[:, :, loo_subj]
+        for v in range(n_vox):
+            ISC[v, loo_subj] = stats.pearsonr(group[v, :],
+                                              subj[v, :])[0]
+    if collapse_subj:
+        ISC = np.mean(ISC, axis=1)
 
-    for p in range(n_perm + 1):
+    for p in range(n_perm):
+        # Randomize phases of D to create next null dataset
+        D = phase_randomize(D, random_state)
         # Loop across choice of leave-one-out subject
+        tmp_ISC = np.zeros((n_vox, n_subj), dtype=float_type)

```

Comment:
In ISFC below, you compute this average iteratively for each subject (so the memory usage doesn't scale with the number of subjects) - you could do the same here, though it doesn't matter much since ISC is much less memory-intensive.